File: euler.scala
object Euler {
    val fibs: Stream[BigInt] = BigInt(0) #:: BigInt(1) #:: fibs.zip(fibs.tail).map(n => n._1 + n._2)

    def gcd(a: Long, b: Long): Long = {
        if (a < b) gcd(b,a)
        else if (b == 0)  a
        else gcd(a-b, b) 
    }

    // least common multiple
    def lcm(a: Long, b: Long)= {
        a*b/gcd(a,b)
    } 

    def sumOfDigit(i: Number): Int = {
       val s = i.toString
       s.map(_.toInt-48).reduce(_+_)
    }
 
    def fact(i: Int): BigInt = {
      if (i == 0) 1
      else if (i==1) 1
      else List.range(2,i+1).map(BigInt(_)).reduce(_*_)
    }

    def __getPrimeFactors(n: BigInt, x: BigInt): List[BigInt] = {
        if (x*x > n) List(n)
        else if (n % x == 0) {
           __getPrimeFactors(n/x, x) :+ x
        }
        else {
           __getPrimeFactors(n, x+(if (x==2) 1 else 2)) 
        }
    }

    def getPrimeFactors(n: BigInt): List[BigInt] = {
        __getPrimeFactors(n, BigInt(2))
    }

    def euler001() = {
      val rg = List.range(1,1000) 
      rg.filter(x => x%3==0 || x%5 ==0).reduce(_+_)
    }
    def euler002() = {
      var a = 1; var b = 1; var c = 2; var sum = 0
      while (c <= 4000000) {
          sum = sum +c
          a = b+c; b = c+a; c=a+b; 
      }
      sum
    }
    def euler003() = {
        getPrimeFactors(BigInt("600851475143")).max
    }
    def euler004() = {
      val a = List.range(100,1000)
      (for { i<- a; j<-a} yield i*j) filter(x=>x.toString==x.toString.reverse) max
    }
    def euler005() = {
      List.range(2L,21L).reduce(lcm(_,_))
    }
    def euler006() = {
      val n = 100
      val l = List.range(1,n+1)
      l.sum * l.sum - l.map(x=>x*x).sum
    }
    def euler008() = {
      val numbers = scala.io.Source.fromFile("data/008.txt").getLines().map(_.trim()).mkString("").map(_.toInt-48)
      (for {i <- List.range(1,numbers.length-4)} yield numbers(i)*numbers(i+1)*numbers(i+2)*numbers(i+3)*numbers(i+4)) max
    }
    def euler013() = {
      val text = scala.io.Source.fromFile("data/13.txt")
      text.getLines().map(BigInt(_)).reduce(_+_).toString().take(10)
    }
    def euler015() =  {
      fact(40)/fact(20)/fact(20)
    }
    def euler016() = {
      sumOfDigit(BigInt(2) pow 1000)
    }
    def euler020() = {
      sumOfDigit(fact(100))
    }
    def euler022() = {
      scala.io.Source.fromFile("data/022.txt").getLines().toList.apply(0).split(",").map(_.stripPrefix("\"").stripSuffix("\"")).sorted.
            zip(Stream.from(1)).map(x=>BigInt(x._1.map(_.toInt-64).sum)*BigInt(x._2)).sum
    }
    def euler025() = {
        fibs.zip(Stream.from(0)) dropWhile (x => x._1.toString.length < 1000) head
    }
    def euler029() = {
      val r = List.range(2,101)
      ((for { a <- r; b <- r} yield BigInt(a).pow(b)) toSet) size
    }
    def euler030() = {
      List.range(2,400000).filter(x=> x == x.toString.map(_.toInt-48).map(x=>x*x*x*x*x).sum).sum
    }
    def euler042() = {
      scala.io.Source.fromFile("data/words.txt").getLines().toList.apply(0).split(',').
            map(x => x.stripPrefix("\"").stripSuffix("\"")).map(x=>(x.map(_.toInt-64) sum)).
            filter(((2 to 30).scanLeft(1)(_+_)).contains(_)).length
    }
    def euler048()= {
      val s = (1 to 1000).map(x=>BigInt(x).pow(x.toInt)).sum.toString
      s.slice(s.length-10, s.length)
    }
    def euler056() = {
      val a = 1 to 100 
      (for {i <- a; j <- a} yield BigInt(i) pow j) map (sumOfDigit(_)) max
    }

    def numNotBouncy(a: Int, e: Int, l: Int) = {
      val d = Math.abs(a-e)
      fact(d+l-2)/fact(l-2)/fact(d)
    }
    def euler113() = {
      (for {a <- List.range(1,10);
            e <- List.range(0,10);
            l <- List.range(2,101)} yield numNotBouncy(a,e,l)).sum + 9
    }
    def euler120() = {
       List.range(3,1001).map(a => List.range(1,a).map(_*2*a%(a*a)).max).sum
    }
    def main(args: Array[String]) {
        var a = System.currentTimeMillis()
        println (euler022())
        println ("time elapsed: " + (System.currentTimeMillis()-a) + " millisec")
    }
}

File: euler.py
import time, math, decimal, itertools
from util import *

def euler002():
    a,b,c,sum = 1,1,2,0
    while (c <= 4000000):
      sum += c
      a = b+c
      b = c+a
      c = a+b
    return sum

def euler003():
    return get_max_prime_divisor(600851475143) 

def euler004():
    return max(filter(lambda x: str(x)==str(x)[::-1], (x*y for x in range(100,1000) for y in range(100, 1000)) ))

def euler005():
    return reduce(lcm, range(1,21))

def euler006():
    n = 100
    l = range(1,n+1)
    return sum(l)*sum(l)-sum(x*x for x in l)

def euler008():
    with open('data/008.txt') as f:
       s = ''.join([x.strip() for x in f.readlines()])
    return max([int(s[i])*int(s[i+1])*int(s[i+2])*int(s[i+3])*int(s[i+4]) for i in range(1,len(s)-4)])

def euler013():
    with open('data/13.txt', 'r') as f:
        lst = f.readlines()
    return str(reduce(lambda x,y: x+y, (int(x) for x in lst)))[:10]

def euler015():
    return math.factorial(40)/math.factorial(20)/math.factorial(20)
    
def euler016():
    return sum_of_digits(2**1000)

def euler018():
    return euler_018_067('18')

def euler_018_067(str_n):
    lst = []
    with open('data/'+str_n+'.txt') as f:
       for l in f.readlines():
           lst.append(map(int, l.split()))
    for x in range(1, len(lst)):
       lst[x][0] = lst[x-1][0]+lst[x][0]
       for y in range(1, len(lst[x])-1):
           lst[x][y] = max(lst[x-1][y-1], lst[x-1][y])+lst[x][y]
       y = len(lst[x])-1
       lst[x][y] = lst[x-1][y-1]+lst[x][y]
    return max(lst[len(lst)-1])

def euler020():
    return sum_of_digits(math.factorial(100))
    
def euler021():
    s = set()
    for n in range(1,10000):
        x = sum(get_all_proper_divisors(n))
        y = sum(get_all_proper_divisors(x))
        if n != x and n == y: s.add(n)
    return sum(s)

def euler023():
    s = set()
    for combi in itertools.combinations_with_replacement(get_abundant_numbers(28123), 2):
        s.add(combi[0]+combi[1])
    t = set(range(1,28124))
    return sum(t.difference(s))

def euler029():
    return len(set(a**b for a in range(2,101) for b in range(2,101)))

def euler030():
    return sum(x for x in range(2, 400000) if x == sum(int(y)**5 for y in str(x)))

def euler037():
    # https://en.wikipedia.org/wiki/List_of_prime_numbers#Two-sided_primes
    two_sided_primes = '23, 37, 53, 73, 313, 317, 373, 797, 3137, 3797, 739397'
    return sum(int(x) for x in two_sided_primes.split(','))

def euler047():
    i = 100
    while(len(get_prime_divisors(i))!=4 or len(get_prime_divisors(i+1))!=4 or len(get_prime_divisors(i+2))!=4 or len(get_prime_divisors(i+3))!=4):
        i += 1
    return i

def euler048():
    return str(sum(map(lambda x: x**x, range(1,1001))))[-10:]

def euler056():
    return max(sum_of_digits(pow(x,y)) for x in range(2,101) for y in range(1,101))

def euler058():
    nums = [3,5,7,9]
    inc  = [10,12,14,16]
    count_prime = 3 
    count_num = 5
    size = 3
    t = 7
    idx = 0
    while 10*count_prime >= count_num:
        idx += 1
        nums = [x+y for x,y in zip(nums, inc)]
        inc  = [x+8 for x in inc]
        count_num += 4
        size += 2
        for x in nums:
           if is_prime(x): count_prime += 1
    print size

def euler067():
    return euler_018_067('67')

def euler080():
    decimal.getcontext().prec=102
    s = 0
    for i in range(2,101):
        if i in [4,9,16,25,36,49,64,81,100]: continue
        t = str(decimal.Decimal(i).sqrt())
        dec = t[2:101]
        s += sum(int(x) for x in dec)+ int(math.sqrt(i)) 
    return s

def euler120():
    return sum(max((2*n*a)%(a*a) for n in range(1,a)) for a in range(3,1001))
        
if __name__ == '__main__' :
    a = time.time()
    print euler047()
    print "time elapsed: %f millisec" % ((time.time()-a)*1000)

